blockchain/vm/internaltx_tracer.go

```diff
@@ -0,0 +1,498 @@
+package vm
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"github.com/klaytn/klaytn/common"
+	"github.com/klaytn/klaytn/common/hexutil"
+	"math/big"
+	"strconv"
+	"time"
+)
+
+var errEvmExecutionReverted = errors.New("evm: execution reverted")
+var errExecutionReverted = errors.New("execution reverted")
+var errInternalFailure = errors.New("internal failure")
+var emptyAddr = common.Address{}
+
+// InternalTxTracer is a full blown transaction tracer that extracts and reports all
+// the internal calls made by a transaction, along with any useful information.
+// It is ported to golang from JS, specifically call_tracer.js
+type InternalTxTracer struct {
+	cfg LogConfig
+
+	callStack []*InternalCall
+	output    []byte
+	err       error
+	errValue  string
+
+	// Below are newly added fields to support call_tracer.js
+	descended        bool
+	revertedContract common.Address
+	ctx              map[string]interface{} // Transaction context gathered throughout execution
+	initialized      bool
+	revertString     string
+}
+
+// NewInternalTxLogger returns a new InternalTxTracer
+func NewInternalTxLogger(cfg *LogConfig) *InternalTxTracer {
+	logger := &InternalTxTracer{
+		callStack: []*InternalCall{},
+		ctx:       map[string]interface{}{},
+	}
+	if cfg != nil {
+		logger.cfg = *cfg
+	}
+	return logger
+}
+
+// InternalCall is emitted to the EVM each cycle and lists information about the current internal state
+// prior to the execution of the statement.
+type InternalCall struct {
+	Type    OpCode         `json:"op"`
+	From    common.Address `json:"from"`
+	To      common.Address `json:"to"`
+	Input   string
+	Gas     uint64 `json:"gas"`
+	GasIn   uint64 `json:"gasIn"`
+	GasUsed uint64 `json:"gasUsed"`
+	GasCost uint64 `json:"gasCost"`
+	Value   string `json:"value"`
+	Err     error  `json:"-"`
+	OutOff  *big.Int
+	OutLen  *big.Int
+	Output  []byte
+
+	calls []*InternalCall
+}
+
+// OpName formats the operand name in a human-readable format.
+func (s *InternalCall) OpName() string {
+	return s.Type.String()
+}
+
+// ErrorString formats the tracerLog's error as a string.
+func (s *InternalCall) ErrorString() string {
+	if s.Err != nil {
+		return s.Err.Error()
+	}
+	return ""
+}
+
+// InternalTxTraceResult is returned data after the end of trace-collecting cycle.
+// It implements an object returned by "result" function at call_tracer.js
+type InternalTxTraceResult struct {
```

Comment:
At 52 line, the struct include `json` tag so that the JS result is a bit different from inner calls and outer calls? How about make them consistent with each other?